### ISSUE TITLE
Update of all yaml instead of just the marked line

### ISSUE THREAD

**Issue by ViBiOh** [fluxcd/image-automation-controller#68]
Hi 👋 

Thank you for your awesome job. I've tested it yesterday and it works like a charm. I've noticed a strange behavior when running automation: every yaml of my repository was rewritten with another formatting.

For exemple, the first commit of `image-automation-controller` on [my repository is this one](https://github.com/vibioh/vith/commit/6914190). The version is [correctly updated](https://github.com/vibioh/vith/commit/6914190#diff-71688e27eea8fddc595d3b07e3e5e72eb4ee1d2c9c0c4cbdb717b07c460f49ffR20) but why the rest also?

I have specific rules for yaml formatting (mainly relying on `prettier`), is there a way (a parameter, a tag) to limit updates to only the lines tagged or only the files which contain tags?

**Comment by pedroparraortega** [comment]
Hi! 
i'm testing the new controllers and when the automation-controller formatted all the manifests in the repository, which i do not think should be the desired behavior, or at least not those which do not have any annotation for the image control.

as example, look at the picture
![Screenshot 2020-12-15 at 10 58 50](https://user-images.githubusercontent.com/7081272/102201067-e831f680-3ec5-11eb-9c6a-b3f3054c4bbe.png)

It is not really easy to handle a configmap like that.

**Comment by pedroparraortega** [comment]
This is getting worst, each time i do a change in my gitRepository where the imageUpdateAutomation lives i get a new commit in the git repository with some new weird formatting changes.

![Screenshot 2020-12-15 at 11 33 17](https://user-images.githubusercontent.com/7081272/102203828-5cba6480-3ec9-11eb-9c36-4670aef2e239.png)
![Screenshot 2020-12-15 at 11 33 24](https://user-images.githubusercontent.com/7081272/102203834-5fb55500-3ec9-11eb-8196-4bfbde89fbab.png)

**Comment by hiddeco** [comment]
Thank you for testing out the alpha controller, and reporting more details around this issue @pedroparraortega / @ViBiOh. :tulip:

Given the severity of the issue you describe, my advise would be to disable the automation for now as it does not seem to properly work (for you) at the moment. As most of the Flux team already is or will soon enjoy some time off, I do not expect this to be worked on before the new year.

Sorry for the inconvenience, and happy early new year wishes, see you in 2021 :fireworks:

**Comment by pedroparraortega** [comment]
> Thank you for testing out the alpha controller, and reporting more details around this issue @pedroparraortega / @ViBiOh. 🌷
> 
> Given the severity of the issue you describe, my advise would be to disable the automation for now as it does not seem to properly work (for you) at the moment. As most of the Flux team already is or will soon enjoy some time off, I do not expect this to be worked on before the new year.
> 
> Sorry for the inconvenience, and happy early new year wishes, see you in 2021 🎆

Yeah, that is what i did. Happy new year to you all, and thanks for the amazing work.

**Comment by javiee** [comment]
I believe the issue I am experiencing could be closely related to this one. I got a gitlab repo containing multiple helm charts. I have placed the marker in a helmRelease resource but I getting an strange behaviour. it would be great if we can somehow restrict down to files or folder where this substitution should happen, 

helmRelease.yml
`image:
        name: docker-registry/plytix-pim
        tag: 1.0.0-beta # {"$imagepolicy": "flux-sytem:plytix-pim-policy:tag"}`

However it looks like it tries to replace helm chart variables

logs:
`{"level":"error","ts":"2020-12-30T11:37:33.572Z","logger":"controller","msg":"Reconciler error","reconcilerGroup":"image.toolkit.fluxcd.io","reconcilerKind":"ImageUpdateAutomation","controller":"imageupdateautomation","name":"plytix-pim","namespace":"flux-system","error":"/tmp/flux-system-flux-system928356935/accounts/templates/deployment.yaml: yaml: line 3: did not find expected key"}`

deployment.yaml

`apiVersion: apps/v1
kind: Deployment
metadata:
  name: {{ .Release.Name }}-gunicorn
`
Happy new year 2021 team!!

**Comment by tommyk** [comment]
i believe this is happening because the use of kustomization/kyaml/kio to pipe ALL the yaml files in git repository to use the setter stuff to edit the image tag.  this explains why the formatting changes even though there is no flags to update an image in your files, and probably why the helm template throws errors as its not valid yaml?

i have been going through the code trying to understand it and fix this issue up as i don't think i can use it until this is fixed.   you can see the line [here](https://github.com/fluxcd/image-automation-controller/blob/8894a7e34d2bba2ec8d41ab9f24e50f4f4299ad5/pkg/update/setters.go#L115) using the kio pipeline.  i was toying around with the `MatchFilesGlob` and realized it seems to only work with exact file names, but not paths.  that certainly wouldn't work if you had an image tag in a `kustomization.yaml` file as you would have many of those and couldn't control it to only go into one directory.
```go
        // get ready with the reader and writer
	reader := &kio.LocalPackageReader{
		PackagePath:        inpath,
		IncludeSubpackages: true,
		// MatchFilesGlob: []string{
		// 	"could.pass.from.crd.and.limit.based.on.filename.but.not.paths.it.seems",
		// },
	}
	writer := &kio.LocalPackageWriter{
		PackagePath: outpath,
	}

	pipeline := kio.Pipeline{
		Inputs:  []kio.Reader{reader},
		Outputs: []kio.Writer{writer},
		Filters: []kio.Filter{kio.FilterAll(&setters2.Set{SetAll: true})},
	}
```

my next thought was to write another "setter" to just find using regex and just replace surgically on  that line only in that file only.   this brings me to my next question from reading the code, you can see there is the idea of "setters" in the `ImageUpdateAutomation` crd, currently the code looks like below and just looks if its set or not, but in the future is the idea that  you could list something like "kio_setter" vs "regex_setter" for different implementations of changing image tags?

code from [here](https://github.com/fluxcd/image-automation-controller/blob/8894a7e34d2bba2ec8d41ab9f24e50f4f4299ad5/controllers/imageupdateautomation_controller.go#L170)
```go
         updateStrat := auto.Spec.Update
	switch {
	case updateStrat.Setters != nil:
		// For setters we first want to compile a list of _all_ the
		// policies in the same namespace (maybe in the future this
		// could be filtered by the automation object).
		var policies imagev1_reflect.ImagePolicyList
		if err := r.List(ctx, &policies, &client.ListOptions{Namespace: req.NamespacedName.Namespace}); err != nil {
			return failWithError(err)
		}

		if err := updateAccordingToSetters(ctx, tmp, policies.Items); err != nil {
			return failWithError(err)
		}
	default:
		log.Info("no update strategy given in the spec")
		// no sense rescheduling until this resource changes
		r.event(auto, events.EventSeverityInfo, "no update strategy in spec, failing trivially")
		imagev1.SetImageUpdateAutomationReadiness(&auto, metav1.ConditionFalse, imagev1.NoStrategyReason, "no update strategy is given for object")
		err := r.Status().Update(ctx, &auto)
		return ctrl.Result{}, err
	}
```

like everyone else, i am excited to help AND use this, and i really appreciate you guys getting the ground work going!  look forward to hearing back, and happy new years everyone!

**Comment by squaremo** [comment]
Erk, yes that's a bit of a disaster. These two things together would mitigate much of the damage:

 - ignore directories that are obviously Helm charts
 - as suggested, only rewrite files that had updates

Those require working around `kio`, which really wants you to trust it to rewrite everything. But I think they are not too intrusive, and there are examples of doing similar things to work from.

**Comment by stefanprodan** [comment]
> ignore directories that are obviously Helm charts

I don't think we should do this, the image patching should work on any Kubernetes YAML no matter if it's in a Helm chart dir or not. 

> as suggested, only rewrite files that had updates

I think this is a first good step, it will still break linters like `prettier` but at least it will not bork all YAMLs in the repo.

**Comment by squaremo** [comment]
> > ignore directories that are obviously Helm charts
> 
> I don't think we should do this, the image patching should work on any Kubernetes YAML no matter if it's in a Helm chart dir or not.

It's pretty unlikely you'd want it to update you chart templates though, isn't it? You might want to update a values.yaml, that's certainly conceivable. So perhaps: ignore `<chart>/templates/`?

An alternative is to ignore .{yaml,yml} files that don't parse; arguably, though, that should be raised as an error.

**Comment by stefanprodan** [comment]
> It's pretty unlikely you'd want it to update you chart templates though, isn't it?

I don't think we should assume what people will do with it, for chart stored in Git, I can see how people would want to patch deployments directly in templates.

**Comment by squaremo** [comment]
> > It's pretty unlikely you'd want it to update you chart templates though, isn't it?
> 
> I don't think we should assume what people will do with it, for chart stored in Git, I can see how people would want to patch deployments directly in templates.

Anything with mustaches `{{ .Values.LikeThis }}` is going to upset the YAML parser, so this causes a dilemma: if a file can't be parsed, should it be ignored, or should it be considered an error? You might accidentally commit invalid YAML to your repo -- you'd want to know if the automation can't process it.

**Comment by stefanprodan** [comment]
> if a file can't be parsed, should it be ignored, or should it be considered an error?

I would log the parse error, issue an event for it so that users get a notification (including the file path) and move to the next file.

**Comment by squaremo** [comment]
> > if a file can't be parsed, should it be ignored, or should it be considered an error?
> 
> I would log the parse error, issue an event for it so that users get a notification (including the file path) and move to the next file.

This is going to annoy people who simply happen to have Helm charts in their repo, and don't expect them to be included in automation in the first place, since they are guaranteed to have error notifications :-/ It will crowd out genuine errors.

**Comment by stefanprodan** [comment]
>  since they are all but guaranteed to have error notifications

Why would they get any notifications if there is no policy marker set in templates?

**Comment by squaremo** [comment]
> > since they are all but guaranteed to have error notifications
> 
> Why would they get any notifications if there is no policy marker set in templates?

We don't know if there's a policy marker or not if we can't parse the file.

**Comment by stefanprodan** [comment]
I think we should filter all yaml files that contain the `$imagepolicy` string and try to parse only those with kio e.g.
```go
if strings.Contains(yamlContent, "$imagepolicy") {
 // try parse 
}
```

**Comment by squaremo** [comment]
> I think we should filter all yaml files that contain the `$imagepolicy` string and try to parse only those with kio

That would also have the effect of cheaply filtering out at least some of the files that don't need to be changed, which is useful. It won't be 100% reliable (because you might have "$imagepolicy" in a file for another reason), but it would err on the side of caution -- no false negatives -- and it would be fairly easy to remedy failures, by e.g., removing the string from the file, or ignoring that file.

**Comment by ViBiOh** [comment]
Instead of parsing the yaml and rewriting it, is it feasible to use Regex? I've quickly tested [this one](https://regex101.com/r/7jjrX3/1) with the example provided in the README. Since `ImageUpdateAutomation` CRD runs with a git reference, we maybe use an equivalent of `git grep` to find appropriate files and use the regex to replace only the part which is relevant for flux.

**Comment by squaremo** [comment]
> is it feasible to use Regex

Using regular expressions for replacing image fields is what I did in the original Flux. It's difficult to make it work in all but simple examples, especially because YAML is structured data. Eventually I wrote a separate library which used a round-tripping YAML parser to do the replacement.

**Comment by jonaskello** [comment]
I still experience this problem and I updated flux yesterday to latest version. The problem I have is that I have a `kustomization.yaml` with  something like this:

```
images:
  - name: XXX
    newName: XXX
    newTag: YYY # {"$imagepolicy": "x:y:tag"}
```
`image-automation-controller` is running image tag `v0.4.0` and it works and updates the `kustomization.yaml` file with a new tag but it also changes all the whitespace for the entire `kustomization.yaml` so in the commit becomes very hard to see what was changed. I use prettier in vscode so it will reformat it back if I save it manually like described in posts above.

My question is if this is supposed to be fixed or not? Reading above posts I got the impression it was fixed but perhaps it was something else that was actually fixed? :-)
